File: web/src/store/council.ts
// One Council run, streamed. Answers arrive interleaved and are keyed by their blind label.
import { create } from "zustand";
import { parseFrames } from "../api/stream";
import type {
  AgreementCell,
  CouncilAnswer,
  CouncilEvent,
  CouncilMember,
  CouncilVerdict,
} from "../api/types";

interface CouncilState {
  running: boolean;
  mode: string;
  detail: string;
  members: CouncilMember[];
  streaming: Record<string, string>;
  answers: Record<string, CouncilAnswer>;
  agreement: AgreementCell[];
  agreementDetail: string;
  verdict: CouncilVerdict | null;
  error: string;

  run: (question: string, modelIds: string[], category: string) => Promise<void>;
  reset: () => void;
}

const EMPTY = {
  mode: "",
  detail: "",
  members: [] as CouncilMember[],
  streaming: {} as Record<string, string>,
  answers: {} as Record<string, CouncilAnswer>,
  agreement: [] as AgreementCell[],
  agreementDetail: "",
  verdict: null,
  error: "",
};

export const useCouncil = create<CouncilState>((set, get) => ({
  running: false,
  ...EMPTY,

  reset: () => set({ ...EMPTY, running: false }),

  run: async (question, modelIds, category) => {
    if (get().running || !question.trim()) return;
    set({ ...EMPTY, running: true });

    let response: Response;
    try {
      response = await fetch("/api/council/run", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ question, model_ids: modelIds, category }),
      });
    } catch (cause) {
      set({ running: false, error: String(cause) });
      return;
    }
    if (!response.ok || !response.body) {
      set({ running: false, error: `${response.status} ${response.statusText}` });
      return;
    }

    const reader = response.body.pipeThrough(new TextDecoderStream()).getReader();
    let buffer = "";
    try {
      for (;;) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += value;
        const { frames, rest } = parseFrames(buffer);
        buffer = rest;
        for (const frame of frames) apply(set, JSON.parse(frame.data) as CouncilEvent);
      }
    } finally {
      set({ running: false });
    }
  },
}));

type Setter = (partial: Partial<CouncilState> | ((s: CouncilState) => Partial<CouncilState>)) => void;

function apply(set: Setter, event: CouncilEvent): void {
  switch (event.type) {
    case "plan":
      set({ members: event.members, mode: event.mode, detail: event.detail });
      break;
    case "answer_token":
      set((s) => ({
        streaming: { ...s.streaming, [event.label]: (s.streaming[event.label] ?? "") + event.text },
      }));
      break;
    case "answer_done":
      set((s) => ({ answers: { ...s.answers, [event.answer.label]: event.answer } }));
      break;
    case "agreement":
      set({ agreement: event.cells, agreementDetail: event.detail });
      break;
    case "verdict":
      set({ verdict: event.verdict });
      break;
    default:
      break;
  }
}

File: server/models/council.py
"""The Council (BRIEF.md 4.6): one prompt, several models, a judge that cannot see the names."""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, Field, RootModel

CouncilMode = Literal["sequential", "mixed"]


class CouncilMember(BaseModel):
    label: str
    model_id: str
    seed: int | None = None
    """Different seeds make "the same model twice" a real comparison rather than a duplicate."""


class CouncilAnswer(BaseModel):
    label: str
    model_id: str
    content: str = ""
    gen_tokens: int = 0
    gen_ms: int = 0
    error: str | None = None


class Ranking(BaseModel):
    label: str
    rank: int
    reason: str = ""


class AgreementCell(BaseModel):
    a: str
    b: str
    similarity: float


class CouncilVerdict(BaseModel):
    ranking: list[Ranking] = []
    synthesis: str = ""
    disagreements: str = ""
    """Where the answers actually differ - the interesting part, not just who won."""
    judge_model_id: str | None = None
    blind: bool = True


class ScoreboardRow(BaseModel):
    model_id: str
    category: str
    wins: int
    appearances: int

    @property
    def win_rate(self) -> float:
        return self.wins / self.appearances if self.appearances else 0.0


class CouncilRequest(BaseModel):
    question: str
    model_ids: list[str] = []
    """Empty means every model this machine can reach."""
    rubric: str = ""
    category: str = "general"
    judge_model_id: str | None = None
    conversation_id: str | None = None


class CouncilReport(BaseModel):
    run_id: str
    question: str
    mode: CouncilMode
    answers: list[CouncilAnswer] = []
    agreement: list[AgreementCell] = []
    agreement_detail: str = ""
    verdict: CouncilVerdict | None = None


# --- streamed events ---


class PlanEvent(BaseModel):
    type: Literal["plan"] = "plan"
    run_id: str
    members: list[CouncilMember]
    mode: CouncilMode
    detail: str


class AnswerStartEvent(BaseModel):
    type: Literal["answer_start"] = "answer_start"
    label: str
    model_id: str


class AnswerTokenEvent(BaseModel):
    type: Literal["answer_token"] = "answer_token"
    label: str
    text: str


class AnswerDoneEvent(BaseModel):
    type: Literal["answer_done"] = "answer_done"
    answer: CouncilAnswer


class AgreementEvent(BaseModel):
    type: Literal["agreement"] = "agreement"
    cells: list[AgreementCell]
    detail: str


class VerdictEvent(BaseModel):
    type: Literal["verdict"] = "verdict"
    verdict: CouncilVerdict


class CouncilDoneEvent(BaseModel):
    type: Literal["council_done"] = "council_done"
    run_id: str


CouncilEvent = Annotated[
    PlanEvent
    | AnswerStartEvent
    | AnswerTokenEvent
    | AnswerDoneEvent
    | AgreementEvent
    | VerdictEvent
    | CouncilDoneEvent,
    Field(discriminator="type"),
]


class CouncilEnvelope(RootModel[CouncilEvent]):
    """Declared as a response model so every event shape reaches the generated TypeScript."""
